File: game/input/actions.lua
-- Edge layer: one key or button press becomes one ActionEvent. Menus and the
-- global toggles consume these. The match screen additionally POLLS the same
-- controls for held state -- see `game.screens.match`.
--
-- Both maps are derived from `game.input.bindings`. Never add a literal key
-- here; add the control there.

local bindings = require("game.input.bindings")

---@alias ActionName
---|"up"
---|"down"
---|"left"
---|"right"
---|"confirm"
---|"back"
---|"pause"
---|"pass_switch"
---|"sprint"
---|"lob"
---|"juke"
---|"equipment"
---|"toggle_fullscreen"
---|"toggle_mute"

---@class ActionEvent
---@field kind "action"
---@field action ActionName
---@field pressed boolean?
---@field source "keyboard"|"gamepad"?

---@class ActionsModule
local actions = {}

---@type table<string, ActionName>
local KEY_MAP = bindings.key_map()

---@type table<string, ActionName>
local GAMEPAD_MAP = bindings.gamepad_map()

---@param name ActionName
---@param pressed boolean?
---@param source "keyboard"|"gamepad"?
---@return ActionEvent
function actions.event(name, pressed, source)
    return {
        kind = "action",
        action = name,
        pressed = pressed == nil or pressed,
        source = source,
    }
end

---@param key string
---@param pressed boolean?
---@return ActionEvent?
function actions.from_key(key, pressed)
    local action = KEY_MAP[key]
    return action and actions.event(action, pressed, "keyboard") or nil
end

-- Equipment shares gamepad B with Back, so which one a press means depends on
-- whether a match is live. The binding table cannot express that; issue #311
-- gives equipment its own button and retires the special case.
---@param button string
---@param pressed boolean?
---@param in_match boolean?
---@return ActionEvent?
function actions.from_gamepad(button, pressed, in_match)
    local action = GAMEPAD_MAP[button]
    local equipment = bindings.control("equipment").buttons[1]
    if button == equipment and in_match then
        action = "equipment"
    end
    return action and actions.event(action, pressed, "gamepad") or nil
end

return actions

File: v2/ts/packages/input/src/actions.ts
// game/input/actions.lua — edge layer: one key or button press becomes one
// ActionEvent. Menus and the global toggles consume these. The (not-yet-
// ported) match screen additionally POLLS the same controls for held state.
//
// Both maps are derived from bindings.ts. Never add a literal key here; add
// the control there.

import { bindings } from "./bindings.ts";

export type ActionName =
  | "up"
  | "down"
  | "left"
  | "right"
  | "confirm"
  | "back"
  | "pause"
  | "pass_switch"
  | "sprint"
  | "lob"
  | "juke"
  | "equipment"
  | "toggle_fullscreen"
  | "toggle_mute";

export type InputSource = "keyboard" | "gamepad";

export interface ActionEvent {
  readonly kind: "action";
  readonly action: ActionName;
  readonly pressed?: boolean;
  readonly source?: InputSource;
}

const KEY_MAP: ReadonlyMap<string, ActionName> = bindings.keyMap();
const GAMEPAD_MAP: ReadonlyMap<string, ActionName> = bindings.gamepadMap();

function event(name: ActionName, pressed = true, source?: InputSource): ActionEvent {
  return {
    kind: "action",
    action: name,
    pressed,
    ...(source !== undefined ? { source } : {}),
  };
}

function fromKey(key: string, pressed?: boolean): ActionEvent | null {
  const action = KEY_MAP.get(key);
  return action !== undefined ? event(action, pressed, "keyboard") : null;
}

// Equipment used to ride gamepad B and shadow Back inside a match. It has
// its own button now, so Back means Back everywhere and this needs no
// context.
function fromGamepad(button: string, pressed?: boolean): ActionEvent | null {
  const action = GAMEPAD_MAP.get(button);
  return action !== undefined ? event(action, pressed, "gamepad") : null;
}

export const actions = { event, fromKey, fromGamepad };
